File: Frost/_src/semantic_version.cpp
#include "../semantic_version.hpp"
#include "Frost.Api/semantic_version.api.hpp"
namespace frost
{
	semantic_version::semantic_version(u16 major, u16 minor, u32 patch, const wchar_t* decoration) :
		major_version(major), minor_version(minor), patch(patch), decoration(decoration) {}
	
	bool semantic_version::check_compatibility(semantic_version target)
	{
		return api::semantic_version::check_compatibility(
			reinterpret_cast<api::semantic_version&>(*this),
			reinterpret_cast<api::semantic_version&>(target));
	}

	semantic_version semantic_version::get_api_version()
	{
		semantic_version result = {};
		api::semantic_version::get_api_version(reinterpret_cast<api::semantic_version*>(&result));
		return result;
	}
}

File: Frost.Net/Models/SemanticVersion.cs
using Frost.Net.Interoperability;
using Frost.Net.Utilities;

namespace Frost.Net.Models;

/// <summary>
/// Semantic version model - Major, Minor, Patch, Decoration
/// </summary>
public sealed class SemanticVersion
{
	public ushort Major { get; set; }
	public ushort Minor { get; set; }
	public uint   Patch { get; set; }
	public string Decoration { get; set; } = string.Empty;



	public bool CheckCompatibility(SemanticVersion target)
	{
		var current_fsv = new FrostApi.SemanticVersion.FrostSemanticVersion()
		{
			major = Major,
			minor = Minor,
			patch = Patch,
			decoration = IntPtr.Zero,
		};

		var target_fsv = new FrostApi.SemanticVersion.FrostSemanticVersion()
		{
			major = target.Major,
			minor = target.Minor,
			patch = target.Patch,
			decoration = IntPtr.Zero,
		};

		return FrostApi.SemanticVersion.CheckCompatibility(current_fsv, target_fsv);
	}



	public static SemanticVersion GetApiVersion()
	{
		FrostApi.SemanticVersion.GetApiVersion(out FrostApi.SemanticVersion.FrostSemanticVersion version);

		return new SemanticVersion()
		{
			Major = version.major,
			Minor = version.minor,
			Patch = version.patch,
			Decoration = Unmanaged.StringFromUnmanagedWstr(version.decoration)
		};
	}
}
